s3/src/main/scala/tamer/s3/package.scala

```diff
@@ -60,124 +56,94 @@ package object s3 {
         .take(paginationMaxPages)
         .timeout(timeoutForFetchAllKeys)
         .runCollect
-        .map(_.toList :+ initialObjListing)
+        .map(_.toList)
       keyList = allObjListings
         .flatMap(objListing => objListing.objectSummaries)
         .map(_.key)
       _                  <- log.debug(s"Current key list has ${keyList.length} elements")
       _                  <- log.debug(s"The first and last elements are ${keyList.sorted.headOption} and ${keyList.sorted.lastOption}")
       previousListOfKeys <- keysR.getAndSet(keyList)
     } yield if (keyList.sorted == previousListOfKeys.sorted) KeysChanged(false) else KeysChanged(true)
-  }
+  }.tap(keysChanged => when(keysChanged.differenceFound)(keysChangedToken.offer(())))
```

Comment:
Every time the list of keys changes (a new file is written or deleted or renamed) it signals it by putting a token in the queue. The queue is bounded to 1 and discards new tokens, this means that if many files are detected multiple times and the other fiber has done nothing, it will still run only once.